### ISSUE TITLE
Improve client error handling

### ISSUE THREAD

**Issue by delucis** [nicolodavis/boardgame.io#723]
This is a proposal for how we could improve error handling in boardgame.io clients.

**Related:** #592 #218

## Current Situation

Errors happen implicitly and there is no way for a client to handle them or even know they happened.

Errors are logged in two key places:

1. The reducer logs errors and returns the original `state` when it encounters them.

2. When in use, the game master does various validation and logs errors too, bailing when it encounters them.

A lot of this validation logic is actually duplicated across master and reducer, which may not be necessary.

## Goals

- A client should know if an error happened and why.

- Errors should be handled in local single-player instances as well as server-connected multiplayer scenarios.

- If possible, the action that triggered an error should be able to receive the error at its call site.

## Core Proposal

1. If the reducer encounters an error, instead of returning the unchanged `state` it should return `{ ...state, error }`.

2. The `error` field should never be persisted — it is a transient side effect of the most recent action specific to a single client.

3. When a master is being used, if there is an error, the `update` should not complete by sending an `update` event with the full state to all clients. Instead, the master should send a new `error` event type to the requesting client. (`error` would be a third event type alongside `update` and `sync`.)

4. The `error` field would have a shape of `{ type, payload? }`, where `type` is one of a range of boardgame.io error codes (see “Error Types” below) and `payload` is additional information where relevant.

5. If possible, client action dispatchers should return a promise that is resolved by the client once the action has been processed. If the action produces an error, the dispatcher would resolve with `{ error }`. See “Errors at call sites” below for more details.

### Error Types

<details>

<summary>A catalogue of potential error types based on current error logging, with suggested error codes.</summary>

| type                         | description                                                |
|:-----------------------------|:-----------------------------------------------------------|
| `update/unauthorized_action` | The action’s `credentials` were missing or invalid         |
| `update/match_not_found`     | The action’s `matchID` was not found                       |
| `action/stale_state_id`      | The action contained a stale state ID                      |
| `action/unavailable_move`    | The requested move is unknown or not currently available   |
| `action/invalid_move`        | The move declared it was invalid (`INVALID_MOVE` constant) |
| `action/inactive_player`     | The player making the action is not currently active       |
| `action/gameover`            | The game has finished                                      |
| `action/action_disabled`     | The requested action is disabled (e.g. undo/redo, events)  |

</details>

### Errors at call sites

<details>
<summary>Could an action dispatcher offer an async response at the call site?</summary>

It would be nice if dispatchers like `client.events.endTurn()` or `client.moves.someMove()` could be awaited so that the call site can do something with the result.

Some more feedback would be welcome here, but would it be possible for the action dispatchers to register callbacks? Internally a dispatcher could do something like:

```js
function dispatcher() {
  // ...
  store.dispatch(/* ... */);

  return new Promise(resolve => {
    client.registerActionCallback(resolve);
  });
}
```

Then when the state updates, the client could check if it was successful or if there was an error, call the callback and remove it.

At the call site users could then do:

```js
client.moves.someMove()
  .then(({ error }) => {
    if (!error) return;
    // handle error
  });
```

There is something unidiomatic about getting an error in `then` instead of in `catch`, but if we avoid throwing errors, users can continue calling dispatchers without worrying about the fact it’s async at all.

There’s an unanswered question here about optimistic updates. Should the promise resolve on a successful optimistic update or must it wait for server response in a multiplayer set-up? This is tricky, because optimistic updates will keep things responsive, but we wouldn’t then be able to resolve with network errors or an error the remote master produced.

</details>

### Invalid moves

<details>

<summary>Supporting arbitrary error data for invalid moves</summary>

There has already been some discussion of letting moves report more data about why they were invalid (see #592). I would suggest the following approach:

- Export an `InvalidMove` factory to be used instead of the `INVALID_MOVE` constant. (We could keep support for the constant for now but deprecate it, only documenting the function.)

    `InvalidMove` would return some predictable object like `{ __error: INVALID_MOVE, payload }`.

- Users could pass arbitrary data to `InvalidMove` and receive that in the reducer’s error payload:

    ```js
    import { InvalidMove } from 'boardgame.io/core';
    const move = (G, ctx) => InvalidMove({ reason: 'no hippos' });
    ```
  
- In the reducer, instead of checking `G === INVALID_MOVE`, we can add a check for the new data structure.

- The above example would produce an `error` in state like:

  ```js
  {
    error: {
      type: 'action/invalid_move',
      payload: { reason: 'no hippos' },
    }
  }
  ```

</details>

### Network errors

Connectivity issues aren’t particularly well managed at the moment because we don’t know for sure when an action times out, errors etc. With the above, the client could implement some timeout/retry/error handling logic.

If an action triggers neither an `update` nor an `error` event after some interval, the client can assume the network hung. It could then roll back optimistic updates or retry.

**Comment by shaoster** [comment]
I'll take a stab at this.

**Comment by shaoster** [comment]
Couple issues I ran into while beginning the spec'd implementation:

1. Given the usual semantics of reducers, I'm not sure the reducer is the best place to emit the action-triggered error:

    > If the reducer encounters an error, instead of returning the unchanged state it should return { ...state, error }.

2. It seems like another way of stating the design goal:

    > If possible, the action that triggered an error should be able to receive the error at its call site.
   
  is that whatever new artifact we emit during a failed action should be correlated to that triggering action.

3. Do you have any architecture doc from which I can try to infer which layer should handle:

   > Errors should be handled in local single-player instances as well as server-connected multiplayer scenarios.

   I see a couple places (most prominently the Master) in the code that can do the aforementioned correlation of action to error, but wanted to make sure I don't start mucking around in a manner that restricts future design space or complicates other features in the roadmap.

**Comment by shaoster** [comment]
# Update

As of #940, the reducer and middleware now expose the data necessary to implement a much richer set of client actions in case of error.

# Call Site API Proposal

## Which Call Sites should be augmented?

I propose first focusing on methods exposed from:

 1. ctx.events
 2. ctx.moves
 3. [Stretch]  ctx.plugins

## Requirements

1. **Backwards Compatibility**: At least in the short term, we should not lock improvements to error handling behind a major version bump of the bgio package. Practically, this means that existing users that depend on the synchronous behavior of the moves/events/plugins methods can keep relying on that synchronicity.
2. The newly introduced error states should be handle-able by the client. This means, at the very minimum, the caller should be able to distinguish between successful invocations and failed invocations.
3. Among failed invocations, the user should be able to distinguish between different failure modes where appropriate. In the case of native bgio events, something like "Not your turn" would be sufficient. In the case of user-implemented moves, a more free form user-populated payload should be communicable.

**Note that the proposals below aren't necessarily mutually exclusive, but rather represent different paradigms to help lay out the design space that I was able to brainstorm.**

## Option 1: Typed Exceptions

Create a distinct, documented, user-handle-able Exception class for each of the error types in [ActionErrorType](https://github.com/boardgameio/boardgame.io/blob/5ac35302ee80626d0bf50a451472bbfc127cf696/src/core/errors.ts#L18) appending payload fields to the exception classes as applicable.

**Benefits**:

1. Flexible framework for exposing pretty much any behavior we'd want to support.
2. Intuitive try/catch error handling patterns for consumers. 

**Detractions**:

1. In all the cases we're handling, the error modes smell much more like "control flow" than "exceptional circumstance." They're deterministic pre-condition violations of some sort or another and should be well-captured by the type system. Exceptions add a wrinkle to otherwise pretty solid types.
2. User code that previously did not pre-validate moves/events, etc... would start failing. While this would be an easy fix on upgrade, it's still breaking and violates the letter, if not spirit of backwards compatibility.

## Option 2: Universal ActionResult return value.

Create a generic ActionResult<Payload, Error> return type that every in-scope call-site will return instead of `void`. Users can decide if and how to handle the result by examining well-documented fields of the returned result object.

**Benefits**:

1. Absolute backwards compatibility. The new return value can be safely ignored.
2. Clean extensibility for non-error payloads as well. This has some pretty concrete use cases for simplifying how the view layer displays the outcome of RNG-dependent moves, like a dice roll or a critical hit.

**Detractions**:

1. Mixing error and non-error states might be okay for user-authored moves, but could be less intuitive for bgio-native events.

## Option 3: Subscriber Callbacks/Promises

The data format of this option would pretty much be identical to that of the synchronous API above. As I see it, using Callbacks vs. Promises are conceptually the same and have the same Pros/Cons with the only meaningful difference being syntax.

The primary difference from Option 2 is that the client method's return is not a post-condition of the reducer having fully processed the requested action; users _must_ await or register callbacks for correctness in many cases.

These two APIs can definitely live alongside one another as async/synchronous variants.

**Benefits**:

1. For experienced developers that are new users of bgio, this is a nice, modern API with very clear semantics that play nice with potentially-expensive moves and UI-responsiveness.

**Detractions**:

1. Promises and callbacks are pretty advanced concepts for new users. Deprecating the old API in favor of an async-only API as proposed is definitely cleaner but could discourage inexperienced developers from adopting bgio in their projects.
2. Obviously, this approach is backwards incompatible. The return of the called client method is no longer a post-condition of the action being processed and G/ctx reflecting resulting changes.

# Personal Recommendation

I think we should combine Options 2 and 3, exposing the option 3 functionality under an `async` namespace under the client interface. For the Option 3 interfaces, I'm also leaning towards promises rather than callbacks, because there's simply better modern language support for the former.

**Comment by delucis** [comment]
Thanks for putting this together @shaoster. I think we’re missing a step here before we tackle the move/event call sites.

The fundamental boardgame.io model currently looks something like this (starting bottom left with the move/event action):

<p align="center"><img src="https://user-images.githubusercontent.com/357379/119719617-4990c500-be69-11eb-9e88-87e883229207.png" alt="Diagram showing the flow of an action processed by client and server." /></p>

When a client calls a move/event, that action is created and the client starts processing it. In the simplest case (no local or socket.io multiplayer) this completes synchronously and the client notifies all subscribers by calling their callback with the new value of `state`. In this synchronous model, returning a result from a dispatcher might be plausible.

However, when multiplayer is in play, things get more complicated:

- A game can set `client: false` on a move to indicate that the move must run on the server (usually because it relies on secret state).
- A plugin can declare that it must be run on the server. For example, if a move uses the randomness API, that will force it to run on the server (also because PRNG state is secret).
- Some events also run on the server exclusively.

So we actually have two update paths (solid blue and dashed red arrows in the diagram):

1. The action is allowed to run locally, so the local reducer completes and notifies subscribers synchronously with the updated state. At some point in the future the server update will also arrive, but will be ignored by the client that updated optimistically.
2. The action is *not* allowed to run locally, so the local reducer does not update state, calling subscribers synchronously with the unchanged state. When the server update is received, the client will call subscribers again with the now updated state.

The implication is that any call site API must be asynchronous (I’d recommend by returning a promise), because we can’t know ahead of time if we can definitely complete the full update cycle synchronously. For case 1, the optimistic update or error could be used to immediately resolve call-site promises. For case 2, the call-site promises should be resolved only once the server update/error is received.

I would propose before working on the call-site API, we should first focus on the existing basic API, which will in any case be a pre-requisite for call-site promises. It will also get the infrastructure set up needed for transmitting the error from the server to the client. I propose updating the client subscription callbacks so they receive `(state, error)` instead of just `(state)`, so a plain JS user of the client can do something like:

```js
// state should be defined even when we provide error to preserve backwards compatibility
client.subscribe((state, error) => {
  if (error) console.error(error.type);
});

client.moves.invalidMove();
// logs 'action/invalid_move'
```

The React client abstracts the plain JS client by spreading state into a board component’s props. I guess we would add the error there directly, so a board might look like:

```js
const Board = ({ moves, error }) => {
  return error
    ? <p>Error: { error.type }</p>
    : <button onClick={() => moves.invalidMove()}>Move</button>;
};
// Clicking the move button causes the paragraph to read “Error: action/invalid_move”
```

I’m also open to a more descriptive name, like `actionError`, which is more important in React because it’s a named prop rather than just a function parameter.

According to this model the parts of closing this issue are:

1. [x] ~~Raise error transients from the reducer~~ (done in #940).
2. [ ] Wire up server, client, and reducer to pass errors to subscribers.
3. [ ] Emit errors from the server for the errors the master raises (not just reducer errors)
4. [ ] Set up a call-site API to allow for action dispatchers to return promises.

What do you think?

**Comment by shaoster** [comment]
Sounds great. I think one pre-requisite that might be worth pursuing before wiring up the server/client/reducer to pass errors, is a way to reconcile the caller semantics between the dashed (async) and solid (sync) paths.

The existing basic API is a somewhat weird place to put this logic for now for a couple reasons:
1.) This is user-exposed, so projects will conceivably be built against these new semantics.
2.) Given that this is a "public" API, there is something a little code-smelly about exposing a state update that isn't really a state update. Namely, the user's code has to maintain _additional_ state to be able to correlate the error with the spawning action in the general case, or alternatively add presentation-layer concerns to the body of the error returned in the move.

Additionally, could you explain why adding an extra argument to the basic API is a pre-requisite to a call-site update? Is this just so the master-routed update has plumbing to get back to the right client in a multiplayer setup?

**Comment by shaoster** [comment]
So my thought process here is that the innerText-setting workflow you describe above actually treats the presence of a transient "error" as a persistent state.

Typically, these transient errors are going to be handled in a manner consistent with 2 properties:
1. Errors don't represent persistent state but rather events. State updates in the UI can easily be derived from these events, but in most cases the majority of the DOM tree won't be touched by an errored client action.

   In React, a typical pattern would probably look more like:
   
   ```javascript
   function Card({cardId}) {
     const { dispatchGlobalError } = useContext(GlobalErrorContext);
     const { G, ctx } = useContext(BgioContext);
     // Card-local state. No need for extra metadata in `error` to correlate back to this component.
     const [maybeError, setMaybeError] = useState(null);
     useEffect(() => {
        if (maybeError) {
          // Any card-specific animations/stateful effects here.
          // ...
          // Equally easy to bubble up global error handling...
          dispatchGlobalError(maybeError);
          setMaybeError(null);
       }
     }, [maybeError, setMaybeError])
     const handleClick = (cardId) => {
       ctx.moves.play(cardId).then({error} => setMaybeError(error))
     }
     return <Button onClick={() => handleClick(cardId)}>Play Card</Button>
   }
   ```

   ```javascript
   function Hand({cardIds}) {
     return <>
       {cardIds.map(cardId => <Card cardId={cardId}/>)}
     </>;
   }
   ```

**Comment by delucis** [comment]
Oh, I’m not saying there should be *no* call site API. I totally understand that for certain implementations it would be preferable to have that way to associate action with error. I just also think there should be a way for code that didn’t dispatch an action to find out there was an error without the need for additional user-implemented infrastructure such as global error contexts or a message bus, which each call site needs to individually remember to tap into. For a hypothetical `ErrorMessage` component in React or plain JS there is no difference whether it gets its value from the boardgame.io client or from some user-land infrastructure — an error just shows up — but for the developer the difference is they have to build that infrastructure and I think we have a pretty clear path to providing a simple API they can use when they don’t need the fine-grained control of your examples.

Writing a component that consumes a declarative error object and displays it when it’s defined is a pretty simple concept to grasp, especially for less advanced users. It’s important to remember that not everyone is comfortable wiring up their own carefully calibrated infrastructure that filters and funnels errors — or that people may want a quick and easy way to access `error` for debugging purposes before building a system to decide what to display, what to log, what to ignore.

My proposal was just to tackle the “simple” API first as a way of getting all the plumbing in place and then add the call site API, but if you want to try this in some other order or all at once, that’s also fine by me 🙂

**Comment by shaoster** [comment]
Great. With that global singleton ErrorMessage use case in mind, I'll make sure that there's at least a plausibly simple tutorial coded up that demonstrates how to wire that up both in React and raw JS.

Additionally, you raised some excellent points in terms of complexity not just being a function of cleanliness user implementation, but rather also being a function of the depth of understanding required by the user to grasp what's going on; it's totally plausible that many of users of bgio (who want error handling) might not have any grasp of how promises, contexts, state, side-effects, or best-practice data models work, or why such complicated things are even talked about in the first place. Maybe they just want to prototype some rules for a game and don't care about any of that abstract mumbo-jumbo.

## TL;DR: I think we should expose a powerful and extensible API that can meet the requirements of advanced users. Simplified/opinionated workflows must be supported as well, but they should be built as a layer ON TOP of that more general API. The rest of this post explains why I believe this.

My concern with exposing the "simple API" for the benefit of unsophisticated programmers (that is, adding the error field to the existing subscriber), is that **as the user becomes more sophisticated and their asks of bgio become better-informed, this part of the API will not grow with them.**

The rough edges in my above examples may become quite frustrating as they start asking questions about key semantics of the extended subscriber interface:

1. When is the subscriber actually being called?

    > Isn't my subscriber only being called when the game state changes? (No, also when there's an error!)
    > Do I actually have to handle the error in my subscriber? What happens if I don't care about errors; will my subscriber be called anyway?

2. What does this "error" object actually represent from a game-flow perspective? Who/what actually raised it?

    > How would I discover the alternate call-site API here, which is a much better fit to answer these questions?
    > If I don't discover this API, might I end up trying to implement this correlation infrastructure myself and give up on bgio?
    > What about non-errored, but nondeterministic outcomes from moves?

3. What am I supposed to do with this error object now that I've passed it to the ErrorMessage component?
    > Can I just blindly pass the incoming error object to a "pure" ErrorMessage magic components like all the other state fields in my subscriber? (No, and it requires a fairly subtle understanding of how a framework like React manages state to understand why).

Based on these concerns (motivated by the limited sample size of my own experience with bgio), I think exposing errors through the existing public subscriber interface would be difficult both to maintain and to reverse.

## ^ All that said, there's another subscriber-based public API that would not bump into any of my above concerns. ##

1. Instead of overloading the capabilities of the existing state-change handler, just add a separate optional handler:

    ```javascript
    export type ErrorHandler = (error: Error) =>  void;
    // subscribe(callback) -> subscribe(stateChangeHandler, errorHandler?: ErrorHandler)
    ```
    This ensures that the existing subscriber is backwards compatible, and gives us a nice place to clearly document the cases where the ErrorHandler's usage is recommended. (e.g. logging, debugging, globally-scoped error-handling).

2. There's a private API internal to bgio that can decorate the error based on the dispatch context. This decorated error would then be passed to the public ErrorHandler above:
    ```javascript
    type PrivateErrorDecorator = (error: Error, errorContext: ErrorContext) => Error;
    ```

3. This lends itself to a pretty clean implementation path:
 
    - [ ] Implement this private error-only handler to set up the client <-> master plumbing without changing any user-visible semantics.
    - [ ] Add the new error subscriber public interface to let people start playing with debug/logging/global error cases, with appropriate (minimal) documentation additions.
    - [ ] Extend the functionality of the private error decorator to implement the  action <-> error correlation mechanisms necessary for the Call-Site API with comprehensive additional tutorials for the "typical" use cases I described above.

**Comment by delucis** [comment]
How do you plan on mapping the separate error handler to React client props? I also considered a separate error handler of the form:

```js
client.subscribe(state => {});
client.subscribeError(error => {}); // or `onError` or whatever
```

But I still think you have to end up mapping `error` to a React prop and for that reason preferred a single callback.

**Comment by shaoster** [comment]
So there's 2 somewhat different simple flows to think about:
1. Raw JS through a subscriber explicitly manipulating the DOM. This doesn't particularly care whether it's one handler or two. An exception is if the user is trying to do fancy synchronization between updates, but that's probably already beyond the scope of the "simple" subscriber API.

2. React through props/components. This is worth diving deeper into.

A core argument I was trying to make in my previous comment is that an error _isn't_ bgio state and we shouldn't encourage beginner programmers to think of the error like a global prop coming from bgio. (It's fundamentally different from ctx and G).

The idiomatic way of handling something like this with a separate error handler would be the following (same idea as from previous comments, but a bit more focused on the subscriber mapping):

```javascript
function Board({client, ...normalStuff}) { /* Or client can come from something global; doesn't really matter */
   const [displayedError, setDisplayedError] = useState(null);
   // Almost straight out of the react hooks tutorial: https://reactjs.org/docs/hooks-effect.html#example-using-hooks-1
   useEffect(() => {
     return client.subscribeError(setDisplayedError)
   }, [client /* if mutable. likely const from the perspective of this component */]);

   // This is basically the pattern for material-ui or react-bootstrap dialog/error components.
   return <>
     <ErrorMessage
          show={displayedError !== null}
          onClose={() => setDisplayedError(null)}
     >
       <ErrorContent error={displayedError}/>
     </ErrorMessage>
     <NormalStuff {...remaining}/>
   </>;
   
  // Alternatively, the user can do something explicitly side-effecting like:
  /*
  useEffect(() => {
    if (displayedError) {
      alert(displayedError.message);
      appendErrorToSomeLog(displayedError);
      setDisplayedError(null);
    }
  }, [displayedError, setDisplayedError]);    
 */
}
```

The crux of the example above is that the `displayedError` explicit react-managed state is the thing that can actually be treated by a beginner programmer as a prop. The bgio-sourced error is an event, not UI state! **We do not want to directly map the error to a prop in the bgio layer!**

The _actual_ error coming through as a prop would likely mislead the beginner developer to write something like this:

```javascript
function Board({error, ...remaining}) {
   if (error) {
     // How would a player dismiss this or a dev animate this away without additional state?
     // Only another client action can clear this.
     return <h1>{error.message}</h1>;
   } else {
      return <NormalStuff {...remaining}/>
   }
}
```

In summary, while bgio mapping the error to a prop is one way for beginner React developers to display errors in their UI, it's a fairly non-idiomatic way that is not much simpler than the "right" way. I'd much prefer to post a tutorial (or just add a built-in debug/sidebar extension) to demonstrate the "right" way to encourage users' growth AND ensure their use of the API can grow with them.

**Comment by delucis** [comment]
Hmm, we certainly don’t want to pass in the whole client directly, but we could theoretically pass an `onError` method from the client for components to use.

A pitfall I see here is that state updates won’t dismiss errors by default and there won’t be a clear way to implement that behaviour in a React component, because the rest of the bgio state arrives as declarative props, so we can’t `subscribe` and clear an error on update. This is a problem with mixing subscription/prop models. Conversely with an `error` prop, both approaches are supported fairly well.

1. Store the error prop and allow it to be manually dismissed (other side effects like in your example code can be implemented similarly and the code is almost identical to using a subscription):

    ```js
    const Board = ({ error }: { error: ErrorInterface | null }) => {
      const [displayedError, setDisplayedError] = useState(error);
      const clearError = () => setDisplayedError(null);
    
      // Keep the displayed error in sync with prop updates:
      useEffect(() => setDisplayedError(error), [error]);
    
      // Or if you want to *only* clear errors manually as in your examples:
      // useEffect(() => {
      //   if (error) setDisplayedError(error)
      // }, [error]);
    
      return displayedError
        ? <button onClick={clearError}>Go away</button>
        : <RestOfBoard />;
    }
    ```

2. Display an error when there is one, hide it once state updates succesfully:

    ```js
    const Board = ({ error }) => {
      return <div>
        {error && <p>Error: {error.message}</p>}
        <RestOfBoard />
      </div>;
    }
    ```

I guess I’m arguing that the React client we offer is an abstraction around the subscription-based plain JS client. It abstracts away any subscription logic by just providing the latest subscription results as props. It is of course valid to prefer subscriptions, in which case one could use the plain JS client more directly and have components subscribe as needed, but mixing these two models seems likely to cause headaches (like the above example, where the lack of a subscription callback on the state side of things makes it hard to update something that also relies on an error subscription callback).

I hope you don’t feel like I’m dismissing the state vs event distinction you make, but unless I’m missing something, the mixing of subscribers and plain props (even if that is along a state/event division) looks like it might be hard to use in practice.

I’m going to ask the other maintainers for some feedback to check if I’m just being boneheaded, but in the meantime, would you like to start with the client–master plumbing? I think we can probably get most if not all of that work done before resolving this implementation detail.

**Comment by shaoster** [comment]
Sounds great to me. The intermediate plumbing layer is unaffected by this discussion.

**Comment by marlonklc** [comment]
I'm looking forward to this feature 😄

**Comment by shaoster** [comment]
Gah. I had some health problems come up and totally dropped the ball on this for like 2 years. I'll try to get caught up and see if any of this still applies/is useful but no promises on timeline.

**Comment by Ma-ve** [comment]
Hi @shaoster ! Has there been any progress on this feature? Am I able to assist in any way?
